apis/oam/v1alpha2/types.go

```diff
@@ -0,0 +1,857 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// A DefinitionReference refers to a CustomResourceDefinition by name.
+type DefinitionReference struct {
+	// Name of the referenced CustomResourceDefinition.
+	Name string `json:"name"`
```

Comment:
No, my intention was that this refers to a CRD by name. For example:

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  group: core.oam.dev
  names:
    kind: ContainerisedWorkload
    listKind: ContainerisedWorkloadList
    plural: containerisedworkloads
    singular: containerisedworkload
  scope: Namespaced
  subresources:
    # etc...
```

We'd refer to the above CRD by name, so:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  definitionRef:
    # Name of the above CRD. We can omit apiVersion and kind here
    # because definition is always kind: CustomResourceDefinition in Kubernetes.
    name: containerisedworkloads.core.oam.dev
```

I believe this makes sense in Kubernetes because when multiple versions of a resource kind exist (e.g. `v1alpha2` and `v1beta1`) they must both be defined by one CRD, and they're both served at the same time. That is, when you create a `v1beta1` `ContainerisedWorkload` an equivalent `v1alpha2` `ContainerisedWorkload` is also served.

If we _really_ wanted to prevent `Component` authors from embedding any version other than `v1beta1`, despite the fact that any other supported versions would implicitly be served by the API server we could instead write something like:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  definitionRef:
    apiVersion: core.oam.dev/v1beta1
    kind: ContainerisedWorkload
```

I prefer to explicitly reference the CRD by name, because it makes it easier for our controllers to find and get the CRD for a particular kind of workload in order to validate components that embed said workload kind.

Reply:
Is this valid?

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  definitionRef:
    name: core.oam.dev/v1beta1.ContainerisedWorkload
```

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  definitionRef:
    name: ContainerisedWorkload.core.oam.dev/v1beta1
```

Reply:
No; those appear to be encodings of type metadata rather than names. If we wanted to use type metadata I would suggest the `apiVersion` and `kind` style reference from my previous comment. Per that comment, I think referring to a CRD by name is preferable.

Reply:
I think we do really need a version, we could define like this:
```
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  definitionRef:
    name: containerisedworkloads.core.oam.dev
    version: v1beta1
```

Reply:
Thanks, Nic. I think we need to include the version as 

- I am not sure how to pinpoint a version without it
- Not all definitions Ref are CRD which includes version
- We don't want to enforce GKV style

How about 

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: containerisedworkloads.core.oam.dev
spec:
  definitionRef:
    name: ContainerisedWorkload.core.oam.dev
   version: v1beta1
```

Reply:
See my comment: https://github.com/oam-dev/spec/pull/316/files#r383564313

I really like to use CRD name here, I just propose we have an optional filed `version` in case multiple CRDs are there.

Reply:
Perhaps you could help me understand why we'd like to pinpoint one version of a kind defined by a CRD. It's not clear to me why we'd want to do this.

Let's say we have a `kind: CustomResourceDefinition` that defines CR `kind: Example`. The CRD defines the schema of two versions; `v1alpha1` and `v1beta1` - note that all versions of a particular GroupKind must be defined via the same CRD. If I author a new `v1beta1` instance of `kind: Example`  a `v1alpha1` compatible representation of that instance is automatically served. That is to say, I can access my new `kind: Example` CR via both the `v1alpha1` and the `v1beta1` API endpoints and schemas. This is discussed at https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects.

So given this, why would we want to bless only one of these versions as a valid workload? Why not allow any valid version? They'll all be translated to one internal version schema and handled by the same controller once they're instantiated anyhow.

Reply:
@negz , we (Alibaba) discussed internally and agreed that we don't need to add version for now.
As for the name of this field, do you think `schemaName` is more informative?

Reply:
> As for the name of this field

Which field are you proposing we rename? `spec.definitionRef` or `spec.definitionRef.name`?

```yaml
spec:
  definitionRef:
    # This violates Kubernetes API conventions for referencing a resource.
    # References must be a strict subset of the ObjectReference type.
    schemaName: foo
```

```yaml
spec:
  # Presuming we remove spec.definitionRef and instead use schemaName
  # like this we'll be compliant with Kubernetes API conventions. I'd prefer
  # definitionName though.
  schemaName: somecrd.example.org
```

Reply:
the first one, nm then.